x/distribution/keeper/delegation.go

```diff
@@ -166,11 +176,20 @@ func (k Keeper) withdrawDelegationRewards(ctx sdk.Context, val stakingtypes.Vali
 
 	// add coins to user account
 	if !coins.IsZero() {
-		withdrawAddr := k.GetDelegatorWithdrawAddr(ctx, del.GetDelegatorAddr())
+		addr := del.GetDelegatorAddr()
+		acc := k.authKeeper.GetAccount(ctx, addr)
+		smartAcc, isSmart := acc.(SmartRewardAccount)
+		withdrawAddr := addr
+		if !isSmart {
+			withdrawAddr = k.GetDelegatorWithdrawAddr(ctx, addr)
+		}
 		err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, withdrawAddr, coins)
 		if err != nil {
 			return nil, err
 		}
+		if isSmart {
+			smartAcc.PostReward(ctx, coins, k.authKeeper, k.bankKeeper, k.stakingKeeper)
```

Comment:
I think this should be an `AfterWithdrawRewards` hook.

Reply:
We also need to prevent the use of a different `withdrawAddr` when we're dealing with a vesting account with clawback. I could have another hook method, `AllowWithdrawAddr(acc authtypes.AccountI) bool` - although this is odd, since there are no other hook methods that return a value or affect execution of the calling module.

Vesting accounts themselves are precedent for having modules modify their behavior based on account type - in its interaction with the `x/bank` module. However, I'm not sure if that's a good precedent to follow.

So I could use the following odd hook interface.  I'm not sure if the resulting code is any simpler or more idiomatic:

```
type interface DistributionHook {
  AllowWithdrawAddr(acc authtypes.AccountI) bool
  AfterReward(ctx sdk.Context, acc authTypes.AccountI, reward sdk.Coins)
}
```

Or I could keep this as-is. Let me know which you'd prefer.

(BTW, why don't the hook-calling modules allow for an arbitrary number of registered hooks? It doesn't seem any harder.)

```
type StakingKeeper struct {
    ...
    hooks []types.StakingHooks
    ...
}

func (k Keeper) BeforeFoo(ctx sdk.Context, addr sdk.AccAddress, foo types.Whatever) {
    for _, hook := range k.hooks {
        hook.BeforeFoo(ctx, addr, foo)
    }
}
```

Reply:
Okay, so I implemented distribution hooks as I described. It was a little tricky getting the wiring correct to tie it all together, but now it's more general, and I think we'll have some other uses for it.